fastly/block_fastly_service_v1_gzip.go

```diff
@@ -118,33 +106,19 @@ func (h *GzipServiceAttributeHandler) Process(d *schema.ResourceData, latestVers
 			Name:           resource["name"].(string),
 		}
 
+		// NOTE: []interface{} is not comparable in Filter function
+		// convert it into string in advance
+		resource["content_types"] = sliceToString(resource["content_types"].([]interface{}))
+		resource["extensions"] = sliceToString(resource["extensions"].([]interface{}))
+
 		// only attempt to update attributes that have changed
 		modified := setDiff.Filter(resource, oldSet)
```

Comment:
Just thought of a possible issue here as well - you're updating `resource[...]` to be a `string`, but `oldSet` will still have it as `[]interface{}` so will probably always be included in `modified`. It's a bit confusing but I don't think it causes any errors because it will just include the existing value in the update every time. 

There's probably two options to rectify that: either just set `resource["content_types"]` etc to `nil` with a comment to explain that it doesn't matter if you redundantly update these values; or you could update `setDiff.Filter` to be more intelligent than just a simple `!=` equality check (i.e. checking the types before comparing, and adding a better check for more complex values like slices)

Reply:
I was thinking of that pattern as well at first. But, I wasn't sure if that's a good idea to check the types for all inputs unconditionally. That seems a bit inefficient processing.

Like, we could make the `Filter` function take optional parameters (e.g., AFOP) and do some custom conversions/diff based on the block type, for instance. This way, `Filter` function is still backward-compatible.

Reply:
How about something like this? Avoids a panic by using `DeepEqual` which checks the types for you, and makes it clear that we're ignoring the resulting `modified` variable for `content_types` and `extensions` by always updating them. I haven't run the whole acceptance test suite to double check but seems like it would work?

```diff
diff --git a/fastly/block_fastly_service_v1_gzip.go b/fastly/block_fastly_service_v1_gzip.go
index 756a3cec..25813411 100644
--- a/fastly/block_fastly_service_v1_gzip.go
+++ b/fastly/block_fastly_service_v1_gzip.go
@@ -104,22 +104,13 @@ func (h *GzipServiceAttributeHandler) Process(d *schema.ResourceData, latestVers
 			ServiceID:      d.Id(),
 			ServiceVersion: latestVersion,
 			Name:           resource["name"].(string),
+			ContentTypes:   gofastly.String(sliceToString(resource["content_types"].([]interface{}))),
+			Extensions:     gofastly.String(sliceToString(resource["extensions"].([]interface{}))),
 		}
 
-		// NOTE: []interface{} is not comparable in Filter function
-		// convert it into string in advance
-		resource["content_types"] = sliceToString(resource["content_types"].([]interface{}))
-		resource["extensions"] = sliceToString(resource["extensions"].([]interface{}))
-
 		// only attempt to update attributes that have changed
 		modified := setDiff.Filter(resource, oldSet)
 
-		if v, ok := modified["content_types"]; ok {
-			opts.ContentTypes = gofastly.String(v.(string))
-		}
-		if v, ok := modified["extensions"]; ok {
-			opts.Extensions = gofastly.String(v.(string))
-		}
 		if v, ok := modified["cache_condition"]; ok {
 			opts.CacheCondition = gofastly.String(v.(string))
 		}
diff --git a/fastly/diff.go b/fastly/diff.go
index c337c065..8151af66 100644
--- a/fastly/diff.go
+++ b/fastly/diff.go
@@ -2,6 +2,7 @@ package fastly
 
 import (
 	"fmt"
+	"reflect"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
@@ -132,7 +133,7 @@ func (h *SetDiff) Filter(modified map[string]interface{}, oldSet *schema.Set) ma
 
 		if m["name"].(string) == modified["name"].(string) {
 			for k, v := range m {
-				if v != modified[k] {
+				if !reflect.DeepEqual(v, modified[k]) {
 					filtered[k] = modified[k]
 				}
 			}
```